apps/prairielearn/src/migrations/20231031163354_client_fingerprints__create.sql

```diff
@@ -0,0 +1,33 @@
+CREATE TABLE
+  client_fingerprints (
+    id BIGSERIAL PRIMARY KEY,
+    user_id BIGINT REFERENCES users (user_id),
+    user_session_id BIGINT REFERENCES user_sessions (id),
+    ip_address INET NOT NULL,
+    user_agent VARCHAR(255),
+    accept_language VARCHAR(255),
+    created_at TIMESTAMPTZ NOT NULL DEFAULT now()
+  );
+
+-- add client_fingerprint_id to page_view_logs
+ALTER TABLE page_view_logs
+ADD COLUMN client_fingerprint_id BIGINT REFERENCES client_fingerprints (id) ON UPDATE CASCADE ON DELETE CASCADE;
```

Comment:
In practice it shouldn't matter because we'll never delete rows from `client_fingerprints`, but just for correctness I think we don't want to CASCADE on DELETEs, because we want to preserve data even without fingerprints.
```suggestion
ADD COLUMN client_fingerprint_id BIGINT REFERENCES client_fingerprints (id) ON UPDATE CASCADE ON DELETE SET NULL;
```
The same applies to the `client_fingerprint_id` columns on other tables below.

Reply:
That makes sense. I have updated all `client_fingerprint_id` columns to SET NULL instead of DELETE.

Reply:
Thanks! Commit, for reference: d20a462